# Week 8: Independent Lab
Author: Jen Estes \
Course: BGEN 632 - Introduction to Python \
Term: Spring 2025 \
Due Date: April 21st, 2025 

This notebook contains code for using advanced functions in `pandas` to query and filter data. This program uses the DataFrame gh_turnover_four_projects.csv., which was created by Dr. Newton and contains information about four projects hosted on GitHub and the contributors to those projects over a period of time.

### Notebook Setup

----

In [5]:
# import modules
import os
import numpy as np
import pandas as pd

In [6]:
# set working directory
os.chdir("/Users/jenestes/Desktop/week8labs/data") 
os.getcwd()  # confirm 

'/Users/jenestes/Desktop/week8labs/data'

### Load Data

In [8]:
# load data and quick inspect
gh_turnover = pd.read_csv("gh_turnover_four_projects.csv")
gh_turnover.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 473 entries, 0 to 472
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   project_id     473 non-null    int64  
 1   domain         473 non-null    object 
 2   language       473 non-null    object 
 3   windows        473 non-null    int64  
 4   window_idx     473 non-null    int64  
 5   num_team       473 non-null    int64  
 6   num_commits    473 non-null    int64  
 7   blau_gender    473 non-null    float64
 8   Gini_gh_ten    473 non-null    float64
 9   user_id        473 non-null    int64  
 10  Core1          473 non-null    int64  
 11  commits        473 non-null    int64  
 12  propCommits    473 non-null    float64
 13  gender         449 non-null    object 
 14  github_tenure  473 non-null    int64  
 15  leavesNextQ    473 non-null    int64  
dtypes: float64(3), int64(10), object(3)
memory usage: 59.3+ KB


### Rename Columns

In [13]:
# renaming columns 
(gh_turnover
 .rename({'domain':'project_domain',
          'language':'project_language',
          'windows':'project_age',
         'window_idx':'quarter',
          'num_team':'team_size',
          'num_commits':'project_commits',
         'blau_gender':'gender_blau',
          'Gini_gh_ten':'gh_tenure_gini',
          'Core1':'core_dev',
         'commits':'user_commits',
          'propCommits':'commit_proportion',
          'github_tenure':'gh_tenure',
         'leavesNextQ':'leaves_next_q'},
         axis = 1,
         inplace = True)
)
gh_turnover # confirm changes

project_id project_domain project_language  project_age  quarter  \
0          1101          OTHER           Python           23       22   
1          1101          OTHER           Python           23       22   
2          1101          OTHER           Python           23       22   
3          1101          OTHER           Python           23       22   
4          1101          OTHER           Python           23       22   
..          ...            ...              ...          ...      ...   
468       24293      DEV_FRAME           Python           24       23   
469       24293      DEV_FRAME           Python           24       24   
470       24293      DEV_FRAME           Python           24       24   
471       24293      DEV_FRAME           Python           24       24   
472       24293      DEV_FRAME           Python           24       24   

     team_size  project_commits  gender_blau  gh_tenure_gini  user_id  \
0           26            22688        0.000        0.422259  1614129   
1           26            22688        0.000        0.422259  1614101   
2           26            22688        0.000        0.422259    94412   
3           26            22688        0.000        0.422259    88459   
4           26            22688        0.000        0.422259   738787   
..         ...              ...          ...             ...      ...   
468         12             1655        0.153        0.292578   938914   
469         12             1224        0.000        0.229926    11648   
470         12             1224        0.000        0.229926   179354   
471         12             1224        0.000        0.229926   438642   
472         12             1224        0.000        0.229926     5268   

     core_dev  user_commits  commit_proportion gender  gh_tenure  \
0           1         22138           0.975758   male        426   
1           0           124           0.005465   male         46   
2           0            53           0.002336   male        941   
3           0            49           0.002160   male       1870   
4           0            45           0.001983   male        781   
..        ...           ...                ...    ...        ...   
468         0             1           0.000604   male        326   
469         0             1           0.000817   male        893   
470         0             1           0.000817   male       1238   
471         0             1           0.000817   male       1582   
472         0             1           0.000817   male       1903   

     leaves_next_q  
0                0  
1                0  
2                1  
3                0  
4                0  
..             ...  
468              1  
469              0  
470              0  
471              0  
472              0  

[473 rows x 16 columns]

### Data Wrangling
This first section calculates the mean and median of `user_commits`. Then, two new DataFrames are created in two different ways to include columns that start with a *p* (i.e., `(^p)`) **or** contain an *g* (i.e., `(g)`) and rows in which `project_commits` is greater than 9,000. The first DataFrame `turnover_new` does this in 2 steps, whereas the second DataFrame `turnover_newer` does this all in one step using piping.

In [16]:
# mean of `user_commits`
(gh_turnover
.agg({
    'user_commits':['mean']
    })
)

user_commits
mean    339.365751

In [18]:
# median of `user_commits`
(gh_turnover
.agg({
    'user_commits':['median']
    })
)

user_commits
median          18.0

In [20]:
# new DataFrame- columns that start with p or have a g
turnover_new = (gh_turnover
             .filter(regex = "(^p|(g))")
                )
turnover_new

project_id project_domain project_language  project_age  project_commits  \
0          1101          OTHER           Python           23            22688   
1          1101          OTHER           Python           23            22688   
2          1101          OTHER           Python           23            22688   
3          1101          OTHER           Python           23            22688   
4          1101          OTHER           Python           23            22688   
..          ...            ...              ...          ...              ...   
468       24293      DEV_FRAME           Python           24             1655   
469       24293      DEV_FRAME           Python           24             1224   
470       24293      DEV_FRAME           Python           24             1224   
471       24293      DEV_FRAME           Python           24             1224   
472       24293      DEV_FRAME           Python           24             1224   

     gender_blau  gh_tenure_gini gender  gh_tenure  
0          0.000        0.422259   male        426  
1          0.000        0.422259   male         46  
2          0.000        0.422259   male        941  
3          0.000        0.422259   male       1870  
4          0.000        0.422259   male        781  
..           ...             ...    ...        ...  
468        0.153        0.292578   male        326  
469        0.000        0.229926   male        893  
470        0.000        0.229926   male       1238  
471        0.000        0.229926   male       1582  
472        0.000        0.229926   male       1903  

[473 rows x 9 columns]

In [22]:
# outputting just the column names for new DataFrame
turnover_new.columns

Index(['project_id', 'project_domain', 'project_language', 'project_age',
       'project_commits', 'gender_blau', 'gh_tenure_gini', 'gender',
       'gh_tenure'],
      dtype='object')

In [24]:
# selecting rows for which `project_commits` is greater than 9,000
(turnover_new
 .query('project_commits > 9000') 
)

project_id project_domain project_language  project_age  project_commits  \
0          1101          OTHER           Python           23            22688   
1          1101          OTHER           Python           23            22688   
2          1101          OTHER           Python           23            22688   
3          1101          OTHER           Python           23            22688   
4          1101          OTHER           Python           23            22688   
..          ...            ...              ...          ...              ...   
116        1101          OTHER           Python           23            14210   
117        1101          OTHER           Python           23            14210   
118        1101          OTHER           Python           23            14210   
119        1101          OTHER           Python           23            14210   
120        1101          OTHER           Python           23            14210   

     gender_blau  gh_tenure_gini gender  gh_tenure  
0            0.0        0.422259   male        426  
1            0.0        0.422259   male         46  
2            0.0        0.422259   male        941  
3            0.0        0.422259   male       1870  
4            0.0        0.422259   male        781  
..           ...             ...    ...        ...  
116          0.0        0.411372   male        495  
117          0.0        0.411372   male        487  
118          0.0        0.411372   male        317  
119          0.0        0.411372   male        418  
120          0.0        0.411372   male        179  

[114 rows x 9 columns]

In [26]:
# piping notation to complete the last two steps together, DataFrame called turnover_newer
turnover_newer = (gh_turnover
                    .filter(regex = "(^p|(g))")
                    .query('project_commits > 9000')
                )
turnover_newer

project_id project_domain project_language  project_age  project_commits  \
0          1101          OTHER           Python           23            22688   
1          1101          OTHER           Python           23            22688   
2          1101          OTHER           Python           23            22688   
3          1101          OTHER           Python           23            22688   
4          1101          OTHER           Python           23            22688   
..          ...            ...              ...          ...              ...   
116        1101          OTHER           Python           23            14210   
117        1101          OTHER           Python           23            14210   
118        1101          OTHER           Python           23            14210   
119        1101          OTHER           Python           23            14210   
120        1101          OTHER           Python           23            14210   

     gender_blau  gh_tenure_gini gender  gh_tenure  
0            0.0        0.422259   male        426  
1            0.0        0.422259   male         46  
2            0.0        0.422259   male        941  
3            0.0        0.422259   male       1870  
4            0.0        0.422259   male        781  
..           ...             ...    ...        ...  
116          0.0        0.411372   male        495  
117          0.0        0.411372   male        487  
118          0.0        0.411372   male        317  
119          0.0        0.411372   male        418  
120          0.0        0.411372   male        179  

[114 rows x 9 columns]

### Final New DataFrame
This new DataFrame includes columns that end with the letter *e* (i.e., `e$`) or contain the letter *s* (i.e., `s`). Additionally, it only includes rows in which `gh_tenure` is greater than or equal to 1,991. 

In [29]:
gh_turnover_final = (gh_turnover
                        .filter(regex = "(e$|(s))")
                        .query('gh_tenure >= 1991')
                            )
gh_turnover_final

project_language  project_age  team_size  project_commits  user_id  \
339             Java           16          2                7    33272   
457           Python           23         20             3070    88459   

     user_commits  gh_tenure  leaves_next_q  
339             6       2006              0  
457             1       2054              0

In [31]:
# mean of `user_commits`
(gh_turnover_final
.agg({
    'user_commits':['mean']
    })
)

user_commits
mean           3.5

In [33]:
# median of `user_commits`
(gh_turnover_final
.agg({
    'user_commits':['median']
    })
)

user_commits
median           3.5

### Comparing the mean and median of `user_commits`

When the data is restricted to rows in which `gh_tenure` is greater than or equal to 1,991, this only leaves 2 rows, which are `user_id` 33272 and 88459. So, when the mean and median of 'user_commits' is calculated, these values are both the same at 3.5. We can compare this to the full dataset, in which the mean of `user_commits` was 339.365751 and median was 18.0. Both of these values are significantly higher compared to the projects that have higher github tenure. Perhaps this suggests that high tenure could be associated with lower user commits, but there is also much larger variability in projects from the full dataset compared to just looking at two projects. There are many other factors that could be influencing this relationship such as team size, project age, etc.  

#### Lab Complete! No References Used 